WordPress/gutenberg
# Add parent navigation support for the navigator component

Builds on top of https://github.com/WordPress/gutenberg/pull/47827 and addresses an issue raised in #47777 

## What and Why?

Parent or "back" navigation doesn't work as we expect it to work in Global Styles and Site editor. The reason is that `goBack` behaves like a browser history, it shows the previous screen that was visible. Sometimes navigation happens programmatically and sometimes the initial screen rendered is not the top level ones (like after reloading the page). For these situations, goBack doesn't work for us. What we really want is a child/parent relationship where the "back" button always goes back to the parent screen.

This PR adds support for that to the Navigator components.

## How?

There are two main ideas here. A user can define a "tree" of screens by relying on the paths: 

  - `/` is the root screen
  - `/something` is a child of root
  - `/:argument` is also a child of root
  - `/something/grandchild` is a child of `/something` if defined, otherwise it fallback to `/`

Basically to define the hierarchy you can split a path on the `/` character and move up until you find a screen that available.

The second idea of this PR is that `goToParent` is equivalent to `goToBack` and should restore the focus in the same way if the path is the previous in location history.

 - I've added a story to storybook to show how this works.

## Testing Instructions

 1- You can open the "nested navigator" story in storybook and check the behavior.


## ✍️ Dev Note

The `Navigator` component from the `@wordpress/components` package has been enhanced with two additional features:

- it can now match named arguments (ie. `/product/:productId`);
- it now offers a way to navigate to the parent screen via the `goToParent` function and the `NavigatorToParentButton` component.

These two features assume that `NavigatorScreen`s are assigned paths that are hierarchical and follow a URL-like scheme, where each path segment is separated by the `/` character.
